File: backend/src/main/java/org/example/coretrack/service/MaterialService.java
package org.example.coretrack.service;

import java.math.BigDecimal;
import java.util.List;

import org.example.coretrack.dto.material.AddMaterialRequest;
import org.example.coretrack.dto.material.AddMaterialResponse;
import org.example.coretrack.dto.material.AllMaterialSearchResponse;
import org.example.coretrack.dto.material.MaterialDetailResponse;
import org.example.coretrack.dto.material.MaterialGroupResponse;
import org.example.coretrack.dto.material.SearchMaterialResponse;
import org.example.coretrack.dto.material.UpdateMaterialRequest;
import org.example.coretrack.dto.material.UpdateMaterialResponse;
import org.example.coretrack.dto.material.ChangeMaterialStatusRequest;
import org.example.coretrack.dto.material.ChangeMaterialStatusResponse;
import org.example.coretrack.dto.material.MaterialStatusTransitionResponse;
import org.example.coretrack.dto.material.MaterialVariantAutoCompleteResponse;
import org.example.coretrack.dto.material.MaterialSupplierResponse;
import org.example.coretrack.dto.material.DeleteMaterialResponse;
import org.example.coretrack.model.auth.User;
import org.example.coretrack.model.material.MaterialVariant;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface MaterialService {
    // to filter the material that meet the criteria (with company context)
    Page<SearchMaterialResponse> findMaterial(
            String search,
            List<String> groupMaterials,
            List<String> statuses,
            Pageable pageable,
            User currentUser);

    // to return all products matching with search (with company context)
    List<AllMaterialSearchResponse> getAllMaterialsForAutocomplete(String search, User currentUser);
    
    // to return all material variants for autocomplete (with company context)
    List<MaterialVariantAutoCompleteResponse> getAllMaterialVariantsForAutocomplete(String search, User currentUser);
    
    // to return all product (with company context)
    Page<SearchMaterialResponse> findAllMaterials(Pageable pageable, User currentUser);

    // to add a material (with company context)
    AddMaterialResponse createMaterial(AddMaterialRequest request, User createdByUser);

    // to rechieve the detailed material info by id (with company context)
    MaterialDetailResponse getMaterialById(Long id, User currentUser); 

    // to update a material (with company context)
    UpdateMaterialResponse updateMaterial(Long id, UpdateMaterialRequest request, User updatedByUser);

    // to get all material groups (with company context)
    List<MaterialGroupResponse> getAllGroupName(User currentUser);

    // to change material status (with company context)
    ChangeMaterialStatusResponse changeMaterialStatus(Long materialId, ChangeMaterialStatusRequest request, User changedByUser);

    // to get available status transitions for a material (with company context)
    MaterialStatusTransitionResponse getAvailableStatusTransitions(Long materialId, User currentUser);

    // to get suppliers by material variant SKU (with company context)
    List<MaterialSupplierResponse> getSuppliersByMaterialVariantSku(String materialVariantSku, User currentUser);

    // to delete a material (soft delete) (with company context)
    DeleteMaterialResponse deleteMaterial(Long id, User deletedByUser);
}

File: frontend/src/services/materialService.ts
// services/materialService.ts
import apiClient from '../lib/axios';
import { PageResponse } from '../types/PageResponse';
import { Material, ProductDetailResponse, MaterialDetailResponse, UpdateMaterialResponse, MaterialQueryParams, MaterialAutoComplete, MaterialGroup } from '../types/material';
import qs from 'qs'; 
import { MaterialVariantAutoComplete } from '../types/material';
import { MaterialSupplierResponse } from '../types/material';


// to fetch the product after searching + filtering
export const getMaterials = async (
  params:  MaterialQueryParams
  
): Promise<PageResponse<Material>> => {
  const defaultParams = {
    page: 0,      
    size: 10,     
    search: '',   
    groupMaterial: [], 
    status: []        
  };
  const mergedParams = { ...defaultParams, ...params };

  console.log('Sending params to /materials/filter', mergedParams);
  const response = await apiClient.get('/materials/filter', { 
    params: mergedParams,
    paramsSerializer: (params) =>
      qs.stringify(params, { arrayFormat: 'repeat' }) 
      // result: ?groupMaterial=A&groupMaterial=B&status=ACTIVE
  });
  return response.data;
};

// to fetch the material for auto complete while searching
export const getAllMaterialsForAutocomplete = async (): Promise<MaterialAutoComplete[]> => {
  const response = await apiClient.get('/materials/all'); 
  return response.data;
};

// Get material variants for autocomplete
export const getAllMaterialVariantsForAutocomplete = async (search?: string): Promise<MaterialVariantAutoComplete[]> => {
  try {
    const params = search ? `?search=${encodeURIComponent(search)}` : '';
    const response = await apiClient.get(`/materials/variants/autocomplete${params}`);
    return response.data;
  } catch (error) {
    console.error('Error fetching material variants for autocomplete:', error);
    throw error;
  }
};

// to fetch the detail product page 
export const getProductById = async (id: number): Promise<ProductDetailResponse> => {
  const res = await apiClient.get<ProductDetailResponse>(`/products/${id}`);
  return res.data;
};

// to fetch material details by ID
export const getMaterialById = async (id: number): Promise<MaterialDetailResponse> => {
  const res = await apiClient.get<MaterialDetailResponse>(`/materials/${id}`);
  return res.data;
};

// to update material
export const updateMaterial = async (id: number, materialData: any): Promise<UpdateMaterialResponse> => {
  const response = await apiClient.put<UpdateMaterialResponse>(`/materials/${id}`, materialData);
  return response.data;
};

// to add new material
export const addMaterial = async (materialData: any): Promise<any> => {
  const response = await apiClient.post('/materials/add-material', materialData);
  return response.data;
};

// to fetch all group 
export const getAllMaterialGroup = async (): Promise<MaterialGroup[]> => {
  const res = await apiClient.get<MaterialGroup[]>('/materials/material-groups');
  return res.data;
};

// to get available status transitions for a material
export const getAvailableStatusTransitions = async (materialId: number): Promise<any> => {
  const response = await apiClient.get(`/materials/${materialId}/status-transitions`);
  return response.data;
};

// to change material status
export const changeMaterialStatus = async (materialId: number, newStatus: string, reason?: string): Promise<any> => {
  const response = await apiClient.put(`/materials/${materialId}/status`, {
    newStatus,
    reason
  });
  return response.data;
};

// Get suppliers by material variant SKU
export const getSuppliersByMaterialVariantSku = async (materialVariantSku: string): Promise<MaterialSupplierResponse[]> => {
  const response = await apiClient.get(`/materials/variants/${materialVariantSku}/suppliers`);
  return response.data;
};

// to delete a material (soft delete)
export const deleteMaterial = async (id: number): Promise<any> => {
  try {
    const response = await apiClient.delete(`/materials/${id}`);
    return response.data;
  } catch (error) {
    throw error;
  }
};